compiler/rustc_span/src/hygiene.rs

```diff
@@ -1315,105 +1361,72 @@ impl<D: Decoder> Decodable<D> for SyntaxContext {
-    let first_hash = hasher.finish();
+    let mut first_hash = hash_expn(&expn_data, &mut ctx);
 
-    let modified = HygieneData::with(|data| {
+    let disambiguator = HygieneData::with(|data| {
         // If this is the first ExpnData with a given hash, then keep our
         // disambiguator at 0 (the default u32 value)
         let disambig = data.expn_data_disambiguators.entry(first_hash).or_default();
-        data.expn_data[expn_id.0 as usize].as_mut().unwrap().disambiguator = *disambig;
+        let disambiguator = *disambig;
         *disambig += 1;
-
-        *disambig != 1
+        disambiguator
     });
 
-    if modified {
+    if disambiguator != 0 {
         debug!("Set disambiguator for {:?} (hash {:?})", expn_id, first_hash);
-        debug!("expn_data = {:?}", expn_id.expn_data());
+        debug!("expn_data = {:?}", expn_data);
+
+        expn_data.disambiguator = disambiguator;
+        first_hash = hash_expn(&expn_data, &mut ctx);
 
         // Verify that the new disambiguator makes the hash unique
         #[cfg(debug_assertions)]
-        {
-            hasher = StableHasher::new();
-            expn_id.expn_data().hash_stable(&mut ctx, &mut hasher);
-            let new_hash: Fingerprint = hasher.finish();
-
-            HygieneData::with(|data| {
-                assert_eq!(
-                    data.expn_data_disambiguators.get(&new_hash),
-                    None,
-                    "Hash collision after disambiguator update!",
-                );
-            });
-        };
+        HygieneData::with(|data| {
+            assert_eq!(
+                data.expn_data_disambiguators.get(&first_hash),
+                None,
+                "Hash collision after disambiguator update!",
+            );
+        });
+    }
+
+    let first_hash = ExpnHash(first_hash);
+    HygieneData::with(|data| {
+        data.expn_data[expn_id.0 as usize].as_mut().unwrap().disambiguator = disambiguator;
+        debug_assert_eq!(data.expn_hash[expn_id.0 as usize].0, Fingerprint::ZERO);
+        data.expn_hash[expn_id.0 as usize] = first_hash;
+        assert!(data.expn_of_hash.insert(first_hash, expn_id).is_none());
+    });
+}
+
+impl<CTX: HashStableContext> HashStable<CTX> for SyntaxContext {
+    fn hash_stable(&self, ctx: &mut CTX, hasher: &mut StableHasher) {
+        const TAG_EXPANSION: u8 = 0;
+        const TAG_NO_EXPANSION: u8 = 1;
+
+        if *self == SyntaxContext::root() {
+            TAG_NO_EXPANSION.hash_stable(ctx, hasher);
+        } else {
+            TAG_EXPANSION.hash_stable(ctx, hasher);
+            let (expn_id, transparency) = self.outer_mark();
+            expn_id.hash_stable(ctx, hasher);
+            transparency.hash_stable(ctx, hasher);
+        }
+    }
+}
+
+impl<CTX: HashStableContext> HashStable<CTX> for ExpnId {
+    fn hash_stable(&self, ctx: &mut CTX, hasher: &mut StableHasher) {
+        let hash = if *self == ExpnId::root() { Fingerprint::ZERO } else { self.expn_hash().0 };
```

Comment:
Is this an optimization?
`self.expn_hash().0` should return `Fingerprint::ZERO` for the root too.